app/lib/active_storage/download_manager.rb

```diff
@@ -0,0 +1,87 @@
+require 'async'
+require 'async/barrier'
+require 'async/http/internet'
+
+class ActiveStorage::DownloadManager
+  include Utils::Retryable
+  DOWNLOAD_MAX_PARALLEL = ENV.fetch('DOWNLOAD_MAX_PARALLEL') { 10 }
+
+  attr_reader :download_to_dir, :errors
+
+  def download_all(attachments:, on_failure:)
+    Async do
+      internet = Async::HTTP::Internet.new
```

Comment:
Apparemment Async fonctionne avec tous les clients HTTP (en tout cas d'après https://brunosutic.com/blog/async-ruby) – donc on n'est pas obligé de tirer la dépendance à async-http si on n'a pas envie.

Apparemment le seul avantage de async-http sur un bête `URI.get`, c'est qu'il gère HTTP/2 (ce qui est légèrement plus rapide justement si on doit initialiser tout un tas de connexions successivement). Mais Typhoeus gère aussi le HTTP/2. Du coup est-ce qu'on ne voudrait pas essayer avec juste Typhoeus pour l'instant (qui est déjà dans le projet), avant de tirer une dépendance supplémentaire ?

Reply:
Je confirme que j'ai reussi a faire de l async avec un bete (et apparemment dangereux URI.open) . vous emballez pas sur le http2, je crois que les differents serveurs upstream ne le gere pas

Reply:
typhoeus c'est mortel (et possiblement plus performant). en revanche j'ai l'impression que la question n'est pas tant de savoir quel lib utiliser sur le DL ; mais quel approche pour l'après. passer a async-ruby me semble être un pari osé, mais réaliste (approche de simon)

aussi j'étais parti sur async-http par naiveté / crainte sur "s'integrer tout l'ecosysteme" (URI.open ; pas moyen que ca fasse de l'IO). une fois connecté a l'object storage d'OVH a faire mes tests ; j'ai eu l'heureuse surprise de tomber sur du http/2. on va requeter le même endpoint ; ca peut valoir le cout de profiter de http/2 ; mais c'est pinailler sur de la perf la jpense.

Reply:
Oué, effectivement quel que soit le backend d'asynchrone on voudra sans doute une méthode qui écrit par chunks plutôt que de tout lire en mémoire d'un coup.

À voir quels helpers sont déjà disponibles ici ou là, ou s'il faut écrire le notre…

Reply:
À la réflexion, je crois que perso je suis assez partant pour ne pas introduire un backend HTTP supplémentaire. Si on n'a pas strictement besoin de `http-async` (parce que `async` + un appel Typhoeus classique fait le job), je serai plutôt pour ne pas introduire de complexité supplémentaire en rajoutant une autre lib.

J'ai l'impression que ça serait essentiellement une affaire de remplacer l'appel `http_client.get` par un `Typhoeus.get`, avec la même gestion de [streamer le fichier sur le disque](https://github.com/typhoeus/typhoeus#streaming-the-response-body). 

Cela dit @LeSim et toi avez déjà bien passé du temps sur cette PR ; du coup si vous préférez laisser `http-async` en l'état, ça me va aussi 👍

Reply:
@LeSim / @kemenaran ; enfait je suis aussi un frugal des gems. Dans le fond, l'implem est dans le gitlog. Si on a vraiment envie / besoin de passer sur de l'async. un git revert le permettra